### PR TITLE
Add conv dequant squash

### PR DESCRIPTION
**Pull request by wozna**
This squash is based on PR https://github.com/PaddlePaddle/Paddle/pull/18754
This squash change chain with conv op. and dequantize op. to conv with forced float output. 
Thanks to this you will avoid additional conversion to int8. 

In mobilenet ssd there are 12 pattern like this and the use of this squash improves accuracy 
(on clx machine):
    - Before squash:  acc_fp32 = 0.7733,  acc_int8=0.77 
    - After that squash: acc_fp32 = 0.7733,  acc_int8=0.7719 

Unfortunately, after a few tests using mkl-dnn and int8 on clx machine, fps decreased:
    - Before squash: avg fps  = 151.634
    - After squash: avg fps = 146.8749
Is's almost 3%. 
That  could be some kind of option if accuracy is more important than performance.

### CODE REVIEW

**Comment by wozna** [comment]
@luotao1  what do you think about the result and option to enable the pass?

**Comment by bingyanghuang** [comment]
Tested on Intel(R) Xeon(R) Gold 6248 CPU @ 2.50GHz

- Test command line
`./paddle/fluid/inference/tests/api/test_analyzer_int8_object_detection "ARGS" --infer_model=third_party/inference_demo/int8v2/mobilenet-ssd/model --infer_data=/home/bingyang/.cache/paddle/dataset/pascalvoc/pascalvoc_full.bin --warmup_batch_size=10 --batch_size=1 --paddle_num_threads=1`

- Before
```
I0821 06:37:47.878208 109578 tester_helper.h:468] --- Performance summary ---
I0821 06:37:47.878234 109578 tester_helper.h:469] FP32: avg fps: 15.5156, avg latency: 64.4513 ms
I0821 06:37:47.878245 109578 tester_helper.h:472] INT8: avg fps: 20.4021, avg latency: 49.0146 ms
I0821 06:37:47.880376 109578 tester_helper.h:452] --- Accuracy summary ---
I0821 06:37:47.880390 109578 tester_helper.h:453] Accepted mAP drop threshold: 0.01. (condition: (FP32_mAP  - INT8_mAP ) <= threshold)
I0821 06:37:47.880398 109578 tester_helper.h:457] FP32: avg mAP 0.7392
I0821 06:37:47.880403 109578 tester_helper.h:459] INT8: avg mAP 0.7287
```

- After
```
I0821 05:35:45.883280 75832 tester_helper.h:468] --- Performance summary ---
I0821 05:35:45.883306 75832 tester_helper.h:469] FP32: avg fps: 15.4869, avg latency: 64.5709 ms
I0821 05:35:45.883317 75832 tester_helper.h:472] INT8: avg fps: 21.2940, avg latency: 46.9617 ms
I0821 05:35:45.885444 75832 tester_helper.h:452] --- Accuracy summary ---
I0821 05:35:45.885458 75832 tester_helper.h:453] Accepted mAP drop threshold: 0.01. (condition: (FP32_mAP  - INT8_mAP ) <= threshold)
I0821 05:35:45.885465 75832 tester_helper.h:457] FP32: avg mAP 0.7392
I0821 05:35:45.885469 75832 tester_helper.h:459] INT8: avg mAP 0.7315
```

- Benchmark conclusion

| Compare | FPS INT8 | Accuracy |
| :---         |     :---:      |          ---: |
| Before   | 20.4021     | 0.7287    |
| After     | 21.2940       | 0.7315      |

**Comment by luotao1** [comment]
@juncaipeng @wzzju Please double-check the https://github.com/PaddlePaddle/Paddle/pull/18905#issuecomment-523284486

File: paddle/fluid/framework/ir/mkldnn/cpu_quantize_squash_pass.h [hunk 1]
```diff
@@ -55,6 +55,11 @@ class CPUQuantizeSquashPass : public FusePassBase {
    */
   void ConvRequantSquash(Graph* graph) const;
 
+  /*
+  *  Squash conv2d with dequant when dequant is the only op after con2d
```
Comment (wojtuss) [inline paddle/fluid/framework/ir/mkldnn/cpu_quantize_squash_pass.h]: `... after conv2d`

File: paddle/fluid/framework/ir/mkldnn/cpu_quantize_squash_pass_tester.cc [hunk 2]
```diff
@@ -57,6 +57,7 @@ void SetOp(ProgramDesc* prog, const std::string& type, const std::string& name,
 // d->Dequant(scale1)->e
 // e->Quant(scale2)->f
 // (f,w2,b2)->Conv2->i
+
```
Comment (wojtuss) [inline paddle/fluid/framework/ir/mkldnn/cpu_quantize_squash_pass_tester.cc]: Please remove the empty line here

File: paddle/fluid/framework/ir/mkldnn/cpu_quantize_squash_pass_tester.cc [hunk 3]
```diff
@@ -85,6 +86,7 @@ static const std::initializer_list<std::string> variable_names{
 // c->Quant1(scale2)->d and d->Conv2->e
 // c->Conv3->f
 // c->Quant2(scale3)->g and g->Conv4->h
+
```
Comment (wojtuss) [inline paddle/fluid/framework/ir/mkldnn/cpu_quantize_squash_pass_tester.cc]: "

File: paddle/fluid/framework/ir/mkldnn/cpu_quantize_squash_pass_tester.cc [hunk 4]
```diff
@@ -192,6 +224,7 @@ void CountNodeTest(const ProgramDesc& prog, int removed_nodes_num) {
 
   int original_nodes_num = graph->Nodes().size();
   RegisterPass(&graph);
+
```
Comment (wojtuss) [inline paddle/fluid/framework/ir/mkldnn/cpu_quantize_squash_pass_tester.cc]: Please keep a consistent approach to inserting empty lines in this and following test methods.
IMO this method can go without empty lines.

File: paddle/fluid/framework/ir/mkldnn/cpu_quantize_squash_pass_tester.cc [hunk 5]
```diff
@@ -161,6 +163,36 @@ ProgramDesc BuildConvMultiRequantProgramDesc(bool use_mkldnn, float scale_out,
...
+  SetOp(&prog, "conv2d", "Conv1", {"a"}, {"b"}, use_mkldnn, scale_out);
+  SetOp(&prog, "dequantize", "Dequant1", {"b"}, {"c"}, use_mkldnn, scale);
+  SetOp(&prog, "concat", "Concat1", {"c"}, {"d"}, use_mkldnn);
+  return prog;
+}
+
+// a->Conv1->b
+// b->Dequant1(Scale1)->c
+// b->Conv2->d
+ProgramDesc BuildConvDequantConvProgramDesc(bool use_mkldnn, float scale_out,
```
Comment (wojtuss) [inline paddle/fluid/framework/ir/mkldnn/cpu_quantize_squash_pass_tester.cc:182]: Here, the `Dequant1`'s output (`c`) is the input to a single `Conv2` only. Probably `Conv1` taking `c` as input is missing here.

Comment (wozna) [reply paddle/fluid/framework/ir/mkldnn/cpu_quantize_squash_pass_tester.cc:182]: The test checks if nothing changes if we have more than one output from Conv1. So Conv1 sholud have two outputs.

Comment (wojtuss) [reply paddle/fluid/framework/ir/mkldnn/cpu_quantize_squash_pass_tester.cc:182]: You are right. Thanks!

**Review by wojtuss — APPROVED** [review]
LGTM

**Comment by wozna** [comment]
@luotao1 if everything is fine then this PR is an improvement ready to merge

**Comment by juncaipeng** [comment]
Tested on Intel(R) Xeon(R) Gold 6271 CPU @ 2.60GHz

* Test command
```
./test_analyzer_int8_object_detection "ARGS" --infer_model=mobilenet-ssd/model/ --infer_data=/dev/shm/pascalvoc_full.bin --warmup_batch_size=10 --batch_size=1 --paddle_num_threads=1
```

* Before
```
I0827 13:02:15.814708 370741 tester_helper.h:471] --- Performance summary ---
I0827 13:02:15.814764 370741 tester_helper.h:472] FP32: avg fps: 12.3974, avg latency: 80.6618 ms
I0827 13:02:15.814790 370741 tester_helper.h:475] INT8: avg fps: 15.9749, avg latency: 62.5980 ms
I0827 13:02:15.817037 370741 tester_helper.h:455] --- Accuracy summary ---
I0827 13:02:15.817065 370741 tester_helper.h:456] Accepted mAP drop threshold: 0.01. (condition: (FP32_mAP  - INT8_mAP ) <= threshold)
I0827 13:02:15.817085 370741 tester_helper.h:460] FP32: avg mAP 0.7392
I0827 13:02:15.817091 370741 tester_helper.h:462] INT8: avg mAP 0.7287
```

* After
```
I0827 11:10:15.486403 161780 tester_helper.h:468] --- Performance summary ---
I0827 11:10:15.486446 161780 tester_helper.h:469] FP32: avg fps: 12.4536, avg latency: 80.2981 ms
I0827 11:10:15.486469 161780 tester_helper.h:472] INT8: avg fps: 16.6916, avg latency: 59.9104 ms
I0827 11:10:15.488651 161780 tester_helper.h:452] --- Accuracy summary ---
I0827 11:10:15.488679 161780 tester_helper.h:453] Accepted mAP drop threshold: 0.01. (condition: (FP32_mAP  - INT8_mAP ) <= threshold)
I0827 11:10:15.488698 161780 tester_helper.h:457] FP32: avg mAP 0.7392
I0827 11:10:15.488703 161780 tester_helper.h:459] INT8: avg mAP 0.7315
```

* Benchmark conclusion

| Compare | FPS INT8 | Accuracy |
| :--: | :--: | :--: |
| Before | 15.9749 | 0.7287 |
| After  | 16.6916  | 0.7315 |

**Pull request merged**